File: src/main/java/net/i2p/crypto/eddsa/math/Representation.java
package net.i2p.crypto.eddsa.math;

/**
 * Available representations for a group element.
 * <ul>
 * <li>P2: Projective representation $(X:Y:Z)$ satisfying $x=X/Z, y=Y/Z$.
 * <li>P3: Extended projective representation $(X:Y:Z:T)$ satisfying $x=X/Z, y=Y/Z, XY=ZT$.
 * <li>P3PrecomputedDouble: P3 but with dblPrecmp populated.
 * <li>P1P1: Completed representation $((X:Z), (Y:T))$ satisfying $x=X/Z, y=Y/T$.
 * <li>PRECOMP: Precomputed representation $(y+x, y-x, 2dxy)$.
 * <li>CACHED: Cached representation $(Y+X, Y-X, Z, 2dT)$
 * </ul>
 */
public enum Representation {
    /**
     * Projective ($P^2$): $(X:Y:Z)$ satisfying $x=X/Z, y=Y/Z$
     */
    P2 {
        public GroupElement toRep(GroupElement groupElement) {
            switch (groupElement.getRepr()) {
                case P2:
                    return new P2GroupElement(groupElement.getCurve(), groupElement.getX(), groupElement.getY(), groupElement.getZ());
                case P3:
                    return new P2GroupElement(groupElement.getCurve(), groupElement.getX(), groupElement.getY(), groupElement.getZ());
                case P1P1:
                    return new P2GroupElement(groupElement.getCurve(), groupElement.getX().multiply(groupElement.getT()), groupElement.getY().multiply(groupElement.getZ()), groupElement.getZ().multiply(groupElement.getT()));
                case PRECOMP:
                    throw new IllegalArgumentException();
                case CACHED:
                    throw new IllegalArgumentException();
                default:
                    throw new UnsupportedOperationException();
            }
        }
    },
    /**
     * Extended ($P^3$): $(X:Y:Z:T)$ satisfying $x=X/Z, y=Y/Z, XY=ZT$
     */
    P3 {
        public GroupElement toRep(GroupElement groupElement) {
            switch (groupElement.getRepr()) {
                case P2:
                    throw new IllegalArgumentException();
                case P3:
                    final Curve curve1 = groupElement.getCurve();
                    final FieldElement x = groupElement.getX();
                    final FieldElement y = groupElement.getY();
                    final FieldElement z = groupElement.getZ();
                    final FieldElement t = groupElement.getT();
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return false ? new P3PreGroupElement(curve1, x, y, z, t) : new P3GroupElement(curve1, x, y, z, t);
                case P1P1:

                    final Curve curve2 = groupElement.getCurve();
                    final FieldElement x1 = groupElement.getX().multiply(groupElement.getT());
                    final FieldElement y1 = groupElement.getY().multiply(groupElement.getZ());
                    final FieldElement z1 = groupElement.getZ().multiply(groupElement.getT());
                    final FieldElement t1 = groupElement.getX().multiply(groupElement.getY());
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return new P3GroupElement(curve2, x1, y1, z1, t1);
                case PRECOMP:
                    throw new IllegalArgumentException();
                case CACHED:
                    throw new IllegalArgumentException();
                default:
                    throw new UnsupportedOperationException();
            }
        }
    },
    /**
     * P3 but also populate dblPrecmp
     */
    P3PrecomputedDouble {
        public GroupElement toRep(GroupElement groupElement) {
            switch (groupElement.getRepr()) {
                case P2:
                    throw new IllegalArgumentException();
                case P3:
                    if (false) {
                        return GroupElement.cached(groupElement.getCurve(), groupElement.getY().add(groupElement.getX()), groupElement.getY().subtract(groupElement.getX()), groupElement.getZ(), groupElement.getT().multiply(groupElement.getCurve().getD2()));
                    } else {
                        throw new IllegalArgumentException();
                    }
                case P1P1:
                    final Curve curve1 = groupElement.getCurve();
                    final FieldElement x = groupElement.getX().multiply(groupElement.getT());
                    final FieldElement y = groupElement.getY().multiply(groupElement.getZ());
                    final FieldElement z = groupElement.getZ().multiply(groupElement.getT());
                    final FieldElement t = groupElement.getX().multiply(groupElement.getY());
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return new P3PreGroupElement(curve1, x, y, z, t);
                case PRECOMP:
                    throw new IllegalArgumentException();
                case CACHED:
                    throw new IllegalArgumentException();
                default:
                    throw new UnsupportedOperationException();
            }
        }
    },
    /**
     * Completed ($P \times P$): $((X:Z),(Y:T))$ satisfying $x=X/Z, y=Y/T$
     */
    P1P1 {
        public GroupElement toRep(GroupElement groupElement) {
            if (groupElement.getRepr() == Representation.P1P1) {
                return GroupElement.p1p1(groupElement.getCurve(), groupElement.getX(), groupElement.getY(), groupElement.getZ(), groupElement.getT());
            }
            if (groupElement.getRepr() == Representation.P2 || groupElement.getRepr() == Representation.P3 || groupElement.getRepr() == Representation.PRECOMP || groupElement.getRepr() == Representation.CACHED)
                throw new IllegalArgumentException();
            throw new UnsupportedOperationException();
        }
    },
    /**
     * Precomputed (Duif): $(y+x,y-x,2dxy)$
     */
    PRECOMP {
        public GroupElement toRep(GroupElement groupElement) {
            if (groupElement.getRepr() == Representation.PRECOMP) {
                return
                        /**
                         * Creates a new group element in PRECOMP representation.
                         *
                         * @param curve The curve.
                         * @param ypx The $y + x$ value.
                         * @param ymx The $y - x$ value.
                         * @param xy2d The $2 * d * x * y$ value.
                         * @return The group element in PRECOMP representation.
                         */new PrecompGroupElement(groupElement.getCurve(), groupElement.getX(), groupElement.getY(), groupElement.getZ());
            }
            if (groupElement.getRepr() == Representation.P2 || groupElement.getRepr() == Representation.P3 || groupElement.getRepr() == Representation.P1P1 || groupElement.getRepr() == Representation.CACHED)
                throw new IllegalArgumentException();
            throw new UnsupportedOperationException();
        }
    },
    /**
     * Cached: $(Y+X,Y-X,Z,2dT)$
     */
    CACHED {
        public GroupElement toRep(GroupElement groupElement) {
            switch (groupElement.getRepr()) {
                case P3:
                    return GroupElement.cached(groupElement.getCurve(), groupElement.getY().add(groupElement.getX()), groupElement.getY().subtract(groupElement.getX()), groupElement.getZ(), groupElement.getT().multiply(groupElement.getCurve().getD2()));
               case P1P1:
                case PRECOMP:
                case P2:
                    throw new IllegalArgumentException();
                case CACHED:
                    return GroupElement.cached(groupElement.getCurve(), groupElement.getX(), groupElement.getY(), groupElement.getZ(), groupElement.getT());
                default:
                    throw new UnsupportedOperationException();
            }
        }
    };

    /**
     * Convert a GroupElement from one Representation to another.
     * TODO-CR: Add additional conversion?
     * $r = p$
     * <p>
     * Supported conversions:
     * <p><ul>
     * <li>P3 $\rightarrow$ P2
     * <li>P3 $\rightarrow$ CACHED (1 multiply, 1 add, 1 subtract)
     * <li>P1P1 $\rightarrow$ P2 (3 multiply)
     * <li>P1P1 $\rightarrow$ P3 (4 multiply)
     *
     * @param groupElement@return A new group element in the given representation.
     */
   abstract public GroupElement toRep(GroupElement groupElement);
}

File: src/main/java/net/i2p/crypto/eddsa/math/Representation.kt
package net.i2p.crypto.eddsa.math

/**
 * Available representations for a group element.
 *
 *  * P2: Projective representation $(X:Y:Z)$ satisfying $x=X/Z, y=Y/Z$.
 *  * P3: Extended projective representation $(X:Y:Z:T)$ satisfying $x=X/Z, y=Y/Z, XY=ZT$.
 *  * P3PrecomputedDouble: P3 but with dblPrecmp populated.
 *  * P1P1: Completed representation $((X:Z), (Y:T))$ satisfying $x=X/Z, y=Y/T$.
 *  * PRECOMP: Precomputed representation $(y+x, y-x, 2dxy)$.
 *  * CACHED: Cached representation $(Y+X, Y-X, Z, 2dT)$
 *
 */
enum class Representation {
    /**
     * Projective ($P^2$): $(X:Y:Z)$ satisfying $x=X/Z, y=Y/Z$
     */
    P2 {
        override fun toRep(groupElement: GroupElement): GroupElement {
            when (groupElement.repr) {
                P2 -> return P2GroupElement(groupElement.curve, groupElement.x, groupElement.y, groupElement.z)
                P3 -> return P2GroupElement(groupElement.curve, groupElement.x, groupElement.y, groupElement.z)
                P1P1 -> return P2GroupElement(groupElement.curve, groupElement.x.multiply(groupElement.t), groupElement.y.multiply(groupElement.z), groupElement.z.multiply(groupElement.t))
                PRECOMP -> throw IllegalArgumentException()
                CACHED -> throw IllegalArgumentException()
                else -> throw UnsupportedOperationException()
            }
        }
    },
    /**
     * Extended ($P^3$): $(X:Y:Z:T)$ satisfying $x=X/Z, y=Y/Z, XY=ZT$
     */
    P3 {
        override fun toRep(groupElement: GroupElement): GroupElement {
            when (groupElement.repr) {
                P2 -> throw IllegalArgumentException()
                P3 -> {
                    val curve1 = groupElement.curve
                    val x = groupElement.x
                    val y = groupElement.y
                    val z = groupElement.z
                    val t = groupElement.t
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return if (false) P3PreGroupElement(curve1, x, y, z, t) else P3GroupElement(curve1, x, y, z, t)
                }
                P1P1 -> {

                    val curve2 = groupElement.curve
                    val x1 = groupElement.x.multiply(groupElement.t)
                    val y1 = groupElement.y.multiply(groupElement.z)
                    val z1 = groupElement.z.multiply(groupElement.t)
                    val t1 = groupElement.x.multiply(groupElement.y)
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return P3GroupElement(curve2, x1, y1, z1, t1)
                }
                PRECOMP -> throw IllegalArgumentException()
                CACHED -> throw IllegalArgumentException()
                else -> throw UnsupportedOperationException()
            }
        }
    },
    /**
     * P3 but also populate dblPrecmp
     */
    P3PrecomputedDouble {
        override fun toRep(groupElement: GroupElement): GroupElement {
            when (groupElement.repr) {
                P2 -> throw IllegalArgumentException()
                P3 -> throw IllegalArgumentException()
                P1P1 -> {
                    val curve1 = groupElement.curve
                    val x = groupElement.x.multiply(groupElement.t)
                    val y = groupElement.y.multiply(groupElement.z)
                    val z = groupElement.z.multiply(groupElement.t)
                    val t = groupElement.x.multiply(groupElement.y)
                    /**
                     * Creates a new group element in P3 representation.
                     *
                     * @param curve The curve.
                     * @param X The $X$ coordinate.
                     * @param Y The $Y$ coordinate.
                     * @param Z The $Z$ coordinate.
                     * @param T The $T$ coordinate.
                     * @param precomputeDoubleOnly If true, populate dblPrecmp, else set to null.
                     * @return The group element in P3 representation.
                     */
                    return P3PreGroupElement(curve1, x, y, z, t)
                }
                PRECOMP -> throw IllegalArgumentException()
                CACHED -> throw IllegalArgumentException()
                else -> throw UnsupportedOperationException()
            }
        }
    },
    /**
     * Completed ($P \times P$): $((X:Z),(Y:T))$ satisfying $x=X/Z, y=Y/T$
     */
    P1P1 {
        override fun toRep(groupElement: GroupElement): GroupElement {
            if (groupElement.repr === Representation.P1P1) {

                /**
                 * Creates a new group element in P1P1 representation.
                 *
                 * @param curve The curve.
                 * @param X The $X$ coordinate.
                 * @param Y The $Y$ coordinate.
                 * @param Z The $Z$ coordinate.
                 * @param T The $T$ coordinate.
                 * @return The group element in P1P1 representation.
                 */
                return P1pGroupElement(groupElement.curve, groupElement.x, groupElement.y, groupElement.z, groupElement.t)
            }
            if (groupElement.repr === Representation.P2 || groupElement.repr === Representation.P3 || groupElement.repr === Representation.PRECOMP || groupElement.repr === Representation.CACHED)
                throw IllegalArgumentException()
            throw UnsupportedOperationException()
        }
    },
    /**
     * Precomputed (Duif): $(y+x,y-x,2dxy)$
     */
    PRECOMP {
        override fun toRep(groupElement: GroupElement): GroupElement {
            if (groupElement.repr === Representation.PRECOMP) {
                return PrecompGroupElement(groupElement.curve, groupElement.x, groupElement.y, groupElement.z)


            }
            if (groupElement.repr === Representation.P2 || groupElement.repr === Representation.P3 || groupElement.repr === Representation.P1P1 || groupElement.repr === Representation.CACHED)
                throw IllegalArgumentException()
            throw UnsupportedOperationException()
        }
    },
    /**
     * Cached: $(Y+X,Y-X,Z,2dT)$
     */
    CACHED {
        override fun toRep(groupElement: GroupElement): GroupElement {
            when (groupElement.repr) {
                P3 -> return CachedGroupElement(groupElement.curve, groupElement.y.add(groupElement.x), groupElement.y.subtract(groupElement.x), groupElement.z, groupElement.t.multiply(groupElement.curve.d2))
                P1P1, PRECOMP, P2 -> throw IllegalArgumentException()
                CACHED -> return CachedGroupElement(groupElement.curve, groupElement.x, groupElement.y, groupElement.z, groupElement.t)
                else -> throw UnsupportedOperationException()
            }
        }
    };

    /**
     * Convert a GroupElement from one Representation to another.
     * TODO-CR: Add additional conversion?
     * $r = p$
     *
     *
     * Supported conversions:
     *
     *
     *  * P3 $\rightarrow$ P2
     *  * P3 $\rightarrow$ CACHED (1 multiply, 1 add, 1 subtract)
     *  * P1P1 $\rightarrow$ P2 (3 multiply)
     *  * P1P1 $\rightarrow$ P3 (4 multiply)
     *
     * @param groupElement@return A new group element in the given representation.
     */
    abstract fun toRep(groupElement: GroupElement): GroupElement
}
